ops/charm.py

```diff
@@ -18,8 +18,12 @@ def restore(self, snapshot):
         if event_action_name != env_action_name:
             # This could only happen if the dev manually emits the action, or from a bug.
             raise RuntimeError('action event kind does not match current action')
-        # Params are loaded at restore rather than __init__ because the model is not available in __init__.
-        self.params = self.framework.model._backend.action_get()
+
+    @property
+    def params(self):
+        # NoneType will be returned if an action has no parameters defined so getattr is used instead of a check for NoneType.
+        self._params = getattr(self, '_params', self.framework.model._backend.action_get())
```

Comment:
This still ends up calling `action-get` on every access, as the parameter value has to be computed before being passed in to `getattr()`.  You could get around this using `hasattr()`:

```python
if not hasattr(self, '_params'):
    self._params = self.model._backend.action_get()
retrun self._params
```

Reply:
Thanks, I'll merge the end result when this lands https://github.com/canonical/operator/pull/132 (it's already addressed there).